File: benchmarks/compute/bench.py
import sys
import time


def fib_recursive(n: int) -> int:
    if n <= 1:
        return n
    return fib_recursive(n - 1) + fib_recursive(n - 2)


def fib_iterative(n: int) -> int:
    MOD = 1_000_000_007
    if n <= 1:
        return n
    a, b = 0, 1
    for _ in range(2, n + 1):
        a, b = b, (a + b) % MOD
    return b


def sieve(n: int) -> int:
    composite = bytearray(n + 1)
    count = 0
    i = 2
    while i <= n:
        if not composite[i]:
            count += 1
            j = i * 2
            while j <= n:
                composite[j] = 1
                j += i
        i += 1
    return count


def sum_loop(n: int) -> int:
    s = 0
    for i in range(1, n + 1):
        s += i
    return s


def main() -> None:
    if len(sys.argv) != 3:
        print("usage: bench.py <test_name> <N>", file=sys.stderr)
        sys.exit(1)

    test = sys.argv[1]
    n = int(sys.argv[2])

    t0 = time.perf_counter_ns()

    if test == "fib_recursive":
        result = fib_recursive(n)
    elif test == "fib_iterative":
        result = fib_iterative(n)
    elif test == "sieve":
        result = sieve(n)
    elif test == "sum":
        result = sum_loop(n)
    else:
        print(f"unknown test: {test}", file=sys.stderr)
        sys.exit(1)

    elapsed = time.perf_counter_ns() - t0

    print(f"time_ns: {elapsed}")
    print(f"result: {result}")


if __name__ == "__main__":
    main()

File: benchmarks/compute/bench.js
'use strict';

function fibRecursive(n) {
    if (n <= 1) return n;
    return fibRecursive(n - 1) + fibRecursive(n - 2);
}

function fibIterative(n) {
    const MOD = 1_000_000_007n;
    if (n <= 1n) return n;
    let a = 0n, b = 1n;
    for (let i = 2n; i <= n; i++) {
        const tmp = (a + b) % MOD;
        a = b;
        b = tmp;
    }
    return b;
}

function sieve(n) {
    const composite = new Uint8Array(n + 1);
    let count = 0;
    for (let i = 2; i <= n; i++) {
        if (!composite[i]) {
            count++;
            for (let j = i * 2; j <= n; j += i) {
                composite[j] = 1;
            }
        }
    }
    return count;
}

function sumLoop(n) {
    let s = 0n;
    for (let i = 1n; i <= n; i++) s += i;
    return s;
}

const args = process.argv;
if (args.length !== 4) {
    process.stderr.write('usage: bench.js <test_name> <N>\n');
    process.exit(1);
}

const test = args[2];
const nRaw = args[3];

let result;
const t0 = process.hrtime.bigint();

switch (test) {
    case 'fib_recursive': {
        const n = parseInt(nRaw, 10);
        result = BigInt(fibRecursive(n));
        break;
    }
    case 'fib_iterative': {
        const n = BigInt(nRaw);
        result = fibIterative(n);
        break;
    }
    case 'sieve': {
        const n = parseInt(nRaw, 10);
        result = BigInt(sieve(n));
        break;
    }
    case 'sum': {
        const n = BigInt(nRaw);
        result = sumLoop(n);
        break;
    }
    default:
        process.stderr.write('unknown test: ' + test + '\n');
        process.exit(1);
}

const elapsed = process.hrtime.bigint() - t0;

process.stdout.write('time_ns: ' + elapsed.toString() + '\n');
process.stdout.write('result: ' + result.toString() + '\n');
